common/lib/transport/connectionmanager.js

```diff
@@ -188,187 +211,94 @@ var ConnectionManager = (function() {
-				 * that would potentially cause a huge spike in load on the load balancer */
-				transportParams.host = Utils.arrPopRandomElement(candidateHosts);
-				self.chooseTransportForHost(transportParams, self.httpTransports.slice(), function(err, httpTransport) {
-					if(err) {
-						if(isFatalOrTokenErr(err)) {
+			if(err) {
+				err = ErrorInfo.fromValues(err);
+				Logger.logAction(Logger.LOG_MINOR, 'ConnectionManager.tryATransport()', 'transport ' + candidate + ' returned err: ' + err.toString());
+
+				/* Comet transport onconnect token errors can be dealt with here.
+				* Websocket ones only happen after the transport claims to be viable,
+				* so are dealt with as non-onconnect token errors */
+				if(Auth.isTokenErr(err)) {
+					/* re-get a token and try again */
+					self.realtime.auth.authorise(null, {force: true}, function(err) {
+						if(err) {
 							callback(err);
 							return;
 						}
-						tryFallbackHosts();
-						return;
-					}
-					/* succeeded */
-					callback(null, httpTransport);
-				});
-			});
-		}
-
-		this.chooseTransportForHost(transportParams, this.httpTransports.slice(), function(err, httpTransport) {
-			if(err) {
-				if(isFatalOrTokenErr(err)) {
-					callback(err);
+						self.tryATransport(transportParams, candidate, callback);
+					});
 					return;
 				}
-				tryFallbackHosts();
+
+				callback(err);
 				return;
 			}
-			/* succeeded */
-			callback(null, httpTransport);
+
+			Logger.logAction(Logger.LOG_MICRO, 'ConnectionManager.chooseTransportForHost()', 'viable transport ' + candidate + '; setting pending');
+			self.setTransportPending(transport, transportParams);
+			callback(null, transport);
 		});
 	};
 
+
 	/**
 	 * Called when a transport is indicated to be viable, and the connectionmanager
 	 * expects to activate this transport as soon as it is connected.
 	 * @param host
-	 * @param transport
+	 * @param transportParams
 	 */
-	ConnectionManager.prototype.setTransportPending = function(transport, mode) {
+	ConnectionManager.prototype.setTransportPending = function(transport, transportParams) {
+		var mode = transportParams.mode;
 		Logger.logAction(Logger.LOG_MINOR, 'ConnectionManager.setTransportPending()', 'transport = ' + transport + '; mode = ' + mode);
 
-		/* this is now pending */
+		Utils.arrDeleteValue(this.proposedTransports, transport);
 		this.pendingTransports.push(transport);
 
 		var self = this;
-		transport.on('connected', function(error, connectionKey, connectionSerial, connectionId, clientId) {
+		transport.once('connected', function(error, connectionKey, connectionSerial, connectionId, clientId) {
 			if(mode == 'upgrade' && self.activeProtocol) {
-				self.scheduleTransportActivation(transport, connectionKey);
+				/*  if ws and xhrs are connecting in parallel, delay xhrs activation to let ws go ahead */
+				if(transport.shortName !== 'web_socket' && Utils.arrIn(self.getUpgradePossibilities(), 'web_socket')) {
```

Comment:
I'd prefer if this literal use of `web_socket` is removed from here. Maybe we define `var OPTIMAL_TRANSPORT = 'web_socket'` at the top of the file? Or it's even taken to be the last element of the transport short names array?